packages/@ember/-internals/meta/lib/meta.ts

```diff
@@ -785,34 +785,27 @@ if (DEBUG) {
     );
 
     let map = this._getOrCreateOwnMap('_values');
-    map[subkey] = value;
+    map[subkey] = value === undefined ? UNDEFINED : value;
   };
 
-  Meta.prototype.peekValues = function(subkey: string) {
-    return this._findInherited2('_values', subkey);
+  Meta.prototype.peekValues = function(key: string) {
+    let val = this._findInherited2('_values', key);
+    return val === UNDEFINED ? undefined : val;
   };
 
-  Meta.prototype.deleteFromValues = function(subkey: string) {
-    delete this._getOrCreateOwnMap('_values')[subkey];
+  Meta.prototype.deleteFromValues = function(key: string) {
+    delete this._getOrCreateOwnMap('_values')[key];
   };
 
-  Meta.prototype.readInheritedValue = function(key, subkey) {
-    let internalKey = `_${key}`;
-
-    let pointer: Meta | null = this;
-
-    while (pointer !== null) {
-      let map = pointer[internalKey];
-      if (map !== undefined) {
-        let value = map[subkey];
-        if (value !== undefined || subkey in map) {
-          return value;
-        }
-      }
-      pointer = pointer.parent;
+  Meta.prototype.readInheritedValue = function(key: string) {
+    let val = this._findInherited2('_values', key);
+    if (val === UNDEFINED) {
+      val = undefined;
+    } else if (val === undefined) {
+      val = UNDEFINED;
```

Comment:
I don't fully understand this branch, why would we return `UNDEFINED` here?

Reply:
I know it is confusing wanted to comment there but now I have better idea improving it, `UNDEFINED` has different meaning in `readInheritedValue` it means it reached the end of inheritance chain of `_values` and have found no value, in `this._findInherited2` `UNDEFINED` means there is `undefined` value found in `_values`, but now I think better way doing it would be removing all the translations above and changing usage of `UNDEFINED` downstream